ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/MessageListenerProxy.java

```diff
@@ -0,0 +1,74 @@
+package org.jboss.narayana.jta.jms;
+
+import com.arjuna.ats.jta.logging.jtaLogger;
+
+import javax.jms.Message;
+import javax.jms.MessageListener;
+import javax.jms.XASession;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import javax.transaction.Transactional;
+import java.lang.reflect.Method;
+
+public class MessageListenerProxy implements MessageListener {
+
+    private MessageListener messageListener;
+    private TransactionManager transactionManager;
+    private XASession xaSession;
+
+    static boolean isDeclaredTransactional(MessageListener messageListener) {
+        try {
+            if (messageListener == null) {
+                return false;
+            }
+
+            Class<? extends MessageListener> listenerClass = messageListener.getClass();
+            if (listenerClass.isAnnotationPresent(Transactional.class)) {
+                return true;
+            } else {
+                Method onMessageMethod = listenerClass.getDeclaredMethod("onMessage", Message.class);
+                return onMessageMethod.isAnnotationPresent(Transactional.class);
+            }
+        } catch (NoSuchMethodException e) {
+            // Should not occur, as messageListener is an instance of javax.jms.MessageListener
+        }
+        return false;
+    }
+
+    public MessageListenerProxy(MessageListener messageListener, XASession session) {
+        this(messageListener, session, com.arjuna.ats.jta.TransactionManager.transactionManager());
+    }
+
+    /**
+     * Mainly for tests
+     */
+    MessageListenerProxy(MessageListener messageListener, XASession session, TransactionManager transactionManager ) {
+        this.messageListener = messageListener;
+        this.transactionManager = transactionManager;
+        this.xaSession = session;
+    }
+
+    @Override
+    public void onMessage(Message jmsMessage) {
```

Comment:
Perfect - thanks Alexis. I did my own research and found this: 
https://github.com/wildfly/wildfly/blob/master/ejb3/src/main/java/org/jboss/as/ejb3/inflow/MessageEndpointInvocationHandler.java#L95
https://github.com/apache/activemq-artemis/blob/master/artemis-ra/src/main/java/org/apache/activemq/artemis/ra/inflow/ActiveMQMessageHandler.java#L301

It seems to suggest that for Artemis it would work unless lines 302-310 were critical (as your PR does not do those lines).

I am concerned that this would only work with certain JMS implementations. Conceptually it does not seem logical to enlist the connection after the message is dequeued, perhaps the JMS/JCA spec mandates that it is possible for the inflow use case. In which case we are back on with this :)